presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTokenSplitManager.java

```diff
@@ -13,131 +13,192 @@
     private final int splitSize;
-    private final IPartitioner partitioner;
 
     @Inject
-    public CassandraTokenSplitManager(CassandraThriftConnectionFactory connectionFactory, @ForCassandra ExecutorService executor, CassandraClientConfig config)
+    public CassandraTokenSplitManager(CassandraSession session, CassandraClientConfig config)
     {
-        this.cassandraThriftClient = new CassandraThriftClient(requireNonNull(connectionFactory, "connectionFactory is null"));
-        this.executor = requireNonNull(executor, "executor is null");
-        this.splitSize = config.getSplitSize();
-        try {
-            this.partitioner = FBUtilities.newPartitioner(config.getPartitioner());
-        }
-        catch (ConfigurationException e) {
-            throw new RuntimeException(e);
-        }
+        this(session, config.getSplitSize());
+    }
+
+    public CassandraTokenSplitManager(CassandraSession session, int splitSize)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.splitSize = splitSize;
     }
 
-    public List<TokenSplit> getSplits(String keyspace, String columnFamily)
-            throws IOException
+    public List<TokenSplit> getSplits(String keyspace, String table)
     {
-        List<TokenRange> masterRangeNodes = cassandraThriftClient.getRangeMap(keyspace);
+        Set<TokenRange> tokenRanges = getTokenRanges();
+
+        if (tokenRanges.isEmpty()) {
+            throw new PrestoException(GENERIC_EXTERNAL, "Cluster metadata is not available. " +
+                    "Please make sure that Cassandra cluster is up and running, " +
+                    "and the contact points specified are correct");
+        }
+
+        if (tokenRanges.stream().anyMatch(TokenRange::isWrappedAround)) {
+            tokenRanges = unwrap(tokenRanges);
+        }
+
+        Optional<TokenRing> tokenRing = createForPartitioner(getPartitioner());
+        long totalPartitionsCount = getTotalPartitionsCount(keyspace, table);
 
-        // canonical ranges, split into pieces, fetching the splits in parallel
         List<TokenSplit> splits = new ArrayList<>();
-        List<Future<List<TokenSplit>>> splitFutures = new ArrayList<>();
-        for (TokenRange range : masterRangeNodes) {
-            // for each range, pick a live owner and ask it to compute bite-sized splits
-            splitFutures.add(executor.submit(new SplitCallable<>(range, keyspace, columnFamily, splitSize, cassandraThriftClient, partitioner)));
+        for (TokenRange tokenRange : tokenRanges) {
+            if (tokenRange.isEmpty()) {
+                continue;
+            }
+            checkState(!tokenRange.isWrappedAround(), "all token ranges must be unwrapped at this step");
+
+            List<String> endpoints = getEndpoints(keyspace, tokenRange);
+            checkState(!endpoints.isEmpty(), "endpoints is empty for token range: %s", tokenRange);
+
+            if (!tokenRing.isPresent()) {
```

Comment:
Basically Cassandra supports 3 types of partitioners.

- Murmur3Partitioner
- RandomPartitioner
- ByteOrderedPartitioner

First 2 of them ensure uniform data distribution across the ring. So, it is easy to slice token ranges based on the total amount of partitions in table. 

ByteOrderedPartitioner, it is a partito ner called `lexicographical` partitioner. This partitioner does't ensure uniform distribution, and it is impossible to predict the number of partition in some particular range. Maybe in future we can make it configurable somehow. So, whoever know his data structure can set how many slices per node does he want. But IRL lexicographical partitioners are not widely used. As far as i understand they are not supported by spark at all. So, i wouldn't bother to much about this now.